### PR TITLE
[Behat] Viewing errors

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Branch?         | 1.2
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

**Comment by lchrusciel** [comment]
Hello Loic!

Thanks a lot for a contribution! I didn't understand it fully, while we were talking about it on slack. But now everything is clear :D 

I may be really beneficial for us, to test 4XX errors, but the current approach doesn't provide a real business value. Don't get me wrong, it is a great start, but I would prefer to see this behavior in a better context. For example, accessing the admin panel after being log out. Or trying to access product page, while it has been deleted. Of course, don't feel forced to test every possible webpage. We may leave it to rest of the contributors. Nevertheless, these generic contexts may help others test their websites better 👍

File: features/error/viewing_errors/viewing_not_found_error.feature [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    As a visitor
+    I need to be able to view a well 404 page when page I requested does not exist
+
+    Background:
+        Given the store operates on a single channel in "United States"
+
+    @ui
+    Scenario: Viewing not found page
+        When I am on not found page
+        Then I should see the title "The page you are looking for does not exist."
```
Comment (lchrusciel) [inline features/error/viewing_errors/viewing_not_found_error.feature]: WDYT about `Then I should be informed that the page does not exist` step instead?

File: features/error/viewing_errors/viewing_forbidden_error.feature [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    As a visitor
+    I need to be able to view a well forbidden page when page I requested is forbidden
+
+    Background:
+        Given the store operates on a single channel in "United States"
+
+    @ui
+    Scenario: Viewing forbidden page
+        When I am on forbidden page
+        Then I should see the title "Unexpected error occurred."
```
Comment (lchrusciel) [inline features/error/viewing_errors/viewing_forbidden_error.feature]: Can we change this behavior, so 403 will explicitly say, that access is forbidden?

Comment (loic425) [reply features/error/viewing_errors/viewing_forbidden_error.feature]: I totally agree, this is a little bug on sylius.

File: src/Sylius/Behat/Page/Shop/Error/ErrorPage.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Page\Shop\Error;
```
Comment (loic425) [inline src/Sylius/Behat/Page/Shop/Error/ErrorPage.php]: @lchrusciel Cause now, it's the general error page, we can use Sylius\Behat\Page\Error namespace. What do you think?

File: features/product/viewing_products/accessing_product_wich_does_not_exist.feature [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    In order to have a good navigation
+    As a visitor
+    I want to be able to be informed that a product does not exits
+
+    Background:
+        Given the store operates on a single channel in "United States"
+
+    @ui
+    Scenario: Accessing a product which does not exist
+        Given the store has a product "T-shirt banana"
```
Comment (lchrusciel) [inline features/product/viewing_products/accessing_product_wich_does_not_exist.feature]: We can remove this line

File: features/product/viewing_products/accessing_product_wich_does_not_exist.feature [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+        When I open page "en_US/products/44 Magnum"
```
Comment (lchrusciel) [inline features/product/viewing_products/accessing_product_wich_does_not_exist.feature]: WDYT about something like `When I try to reach unexistent product`? Behind this step we can just call `$this->showPage->open(['slug' => 'unexisten_product', '_locale' => $localeCode]);`

File: src/Sylius/Behat/Page/Shop/Error/ErrorPage.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+use Sylius\Behat\Page\Page;
+
+class ErrorPage extends Page implements ErrorPageInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getUrl(array $urlParameters = [])
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/Error/ErrorPage.php]: missing return type

File: src/Sylius/Behat/Page/Shop/Error/ErrorPage.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
+<?php
+
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/Error/ErrorPage.php]: May we start adding strict types here?

File: src/Sylius/Behat/Page/Shop/Error/ErrorPage.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    public function getTitle(): string
+    {
+        return $this->getElement('title')->getText();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getDefinedElements()
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/Error/ErrorPage.php]: Missing return type hint

File: src/Sylius/Behat/Page/Shop/Error/ErrorPage.php (hunk 3 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Page/Shop/Error/ErrorPage.php]: Good idea

**Comment by lchrusciel** [comment]
Anyway, @pamil is testing something on repo, so I will merge it next week probably

File: src/Sylius/Behat/Page/Shop/Error/ErrorPage.php (hunk 6 above)
Comment (loic425) [reply src/Sylius/Behat/Page/Shop/Error/ErrorPage.php]: It's an abstract on Sylius\Behat\Page
`abstract protected function getUrl(array $urlParameters = [])`
But this is a protected method, I need to change that

File: src/Sylius/Behat/Page/Shop/Error/ErrorPage.php (hunk 8 above)
Comment (loic425) [reply src/Sylius/Behat/Page/Shop/Error/ErrorPage.php]: Also on SyliusBehatPage
```
protected function getDefinedElements()
{
    return [];
}
```

**Comment by loic425** [comment]
@lchrusciel I think this pull request name could be improved before merging it.
It has no sense now imho. What do tout think ?

**Comment by loic425** [comment]
@pamil can you restart build on 4.1, it crashed.

**Comment by loic425** [comment]
@pamil @lchrusciel 
It failed again on symfony 4.1. But on my travis repo it's green
https://travis-ci.org/loic425/Sylius/jobs/458461014

File: features/admin/redirecting_on_login_after_being_logout.feature [hunk 9]
```diff
@@ -0,0 +1,14 @@
+@administrator_login
+Feature: Redirecting on login page
+    In order to access on my admin dashboard
+    As a logged out Administrator
+    I need to be redirected on login page
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And I have been logged out from administration
```
Comment (lchrusciel) [inline features/admin/redirecting_on_login_after_being_logout.feature]: Maybe we may add being log in in before this step?

File: src/Sylius/Behat/Context/Ui/Admin/DashboardContext.php [hunk 10]
```diff
@@ -29,11 +30,14 @@ public function __construct(DashboardPageInterface $dashboardPage)
     }
 
     /**
-     * @When I open administration dashboard
+     * @When I (try to )open administration dashboard
      */
     public function iOpenAdministrationDashboard()
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Admin/DashboardContext.php]: WDYT about? 
```php
    /**
     * @When I open administration dashboard
     */
    public function iOpenAdministrationDashboard()
    {
        $this->dashboardPage->open();
    }
    /**
     * @When I try to open administration dashboard
     */
    public function iTryToOpenAdministrationDashboard()
    {
        try {
            $this->iOpenAdministrationDashboard();
        } catch (UnexpectedPageException $e) {
        }
    }
```
So we can decide, if we want to catch error or not

**Comment by lchrusciel** [comment]
Thank you, Loïc! :tada:

**Comment by lchrusciel** [comment]
I have also counted it as bug fix due to https://github.com/Sylius/Sylius/pull/9860#discussion_r228429739

File: src/Sylius/Behat/Context/Ui/Admin/DashboardContext.php (hunk 10 above)
Comment (loic425) [reply src/Sylius/Behat/Context/Ui/Admin/DashboardContext.php]: Yep and moreover i've recently seen there is a tryToOpen method on page object.

**Pull request merged**